File: app/src/main/java/com/example/h_mal/shift_tracker/ShiftsCursorAdapter.java
package com.example.h_mal.shift_tracker;

import android.content.ContentUris;
import android.content.Context;
import android.content.DialogInterface;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.FragmentTransaction;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CursorAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.example.h_mal.shift_tracker.Data.ShiftProvider;
import com.example.h_mal.shift_tracker.Data.ShiftsContract.ShiftsEntry;

/**
 * Created by h_mal on 26/12/2017.
 */

public class ShiftsCursorAdapter extends CursorAdapter {

    private final MainActivity activity;

    private Context mContext;

    ShiftProvider shiftProvider;

    public ShiftsCursorAdapter(MainActivity context, Cursor c) {
        super(context, c, 0);
        this.activity = context;
    }


    @Override
    public View newView(Context context, Cursor cursor, ViewGroup parent) {
        return LayoutInflater.from(context).inflate(R.layout.list_item_1, parent, false);
    }



    @Override
    public void bindView(View view, Context context, Cursor cursor) {
        mContext = context;

        TextView descriptionTextView = (TextView) view.findViewById(R.id.location);
        TextView dateTextView = (TextView) view.findViewById(R.id.date);
        TextView totalPay = (TextView) view.findViewById(R.id.total_pay);
        TextView hoursView = (TextView) view.findViewById(R.id.hours);
        TextView h = (TextView) view.findViewById(R.id.h);
        TextView minutesView = (TextView) view.findViewById(R.id.minutes);
        TextView m = (TextView) view.findViewById(R.id.m);
        ImageView editView = (ImageView) view.findViewById(R.id.imageView);

        h.setText("h");
        m.setText("m");

        final String typeColumnIndex = cursor.getString(cursor.getColumnIndexOrThrow(ShiftsEntry.COLUMN_SHIFT_TYPE));
        final String descriptionColumnIndex = cursor.getString(cursor.getColumnIndexOrThrow(ShiftsEntry.COLUMN_SHIFT_DESCRIPTION));
        final String dateColumnIndex = cursor.getString(cursor.getColumnIndexOrThrow(ShiftsEntry.COLUMN_SHIFT_DATE));
        final Float durationColumnIndex = cursor.getFloat(cursor.getColumnIndexOrThrow(ShiftsEntry.COLUMN_SHIFT_DURATION));
        final Float unitsColumnIndex = cursor.getFloat(cursor.getColumnIndexOrThrow(ShiftsEntry.COLUMN_SHIFT_UNIT));
        final Float totalpayColumnIndex = cursor.getFloat(cursor.getColumnIndexOrThrow(ShiftsEntry.COLUMN_SHIFT_TOTALPAY));

        descriptionTextView.setText(descriptionColumnIndex);
        dateTextView.setText(newDate(dateColumnIndex));
        totalPay.setText(String.format("%.2f",totalpayColumnIndex));

        if (typeColumnIndex.equals("Piece Rate") && durationColumnIndex == 0){
            hoursView.setText(String.valueOf(unitsColumnIndex));
            h.setText("");
            minutesView.setText("");
            m.setText("pcs");
        }else
//            if(typeColumnIndex.equals("Hourly") || typeColumnIndex.equals("hourly"))
        {
            hoursView.setText(timeValues(durationColumnIndex)[0]);
            minutesView.setText(timeValues(durationColumnIndex)[1]);
        }

        final long ID = cursor.getLong(cursor.getColumnIndexOrThrow(ShiftsEntry._ID));
        final Uri currentProductUri = ContentUris.withAppendedId(ShiftsEntry.CONTENT_URI, ID);

        final Bundle b = new Bundle();
        b.putString("uri",String.valueOf(currentProductUri));

        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
//                activity.clickOnViewItem(ID);
                FragmentTransaction fragmentTransaction = (MainActivity.fragmentManager).beginTransaction();
                FurtherInfoFragment fragment2 = new FurtherInfoFragment();
                fragment2.setArguments(b);
                fragmentTransaction.replace(R.id.container,fragment2).addToBackStack("furtherinfo").commit();
            }
        });

        editView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                FragmentTransaction fragmentTransaction = (MainActivity.fragmentManager).beginTransaction();
                FragmentAddItem fragment3 = new FragmentAddItem();
                fragment3.setArguments(b);
                fragmentTransaction.replace(R.id.container,fragment3).addToBackStack("additem").commit();
            }
        });

        view.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                System.out.println("long click: " + ID);

                android.app.AlertDialog.Builder builder = new android.app.AlertDialog.Builder(mContext);
                builder.setMessage("Are you sure you want to delete");
                builder.setPositiveButton("delete", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        deleteProduct(ID);
                    }
                });
                builder.setNegativeButton("cancel", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {

                        if (dialog != null) {
                            dialog.dismiss();
                        }
                    }
                });

                android.app.AlertDialog alertDialog = builder.create();
                alertDialog.show();
                return true;
            }
        });

    }

    private void deleteProduct(Long id) {
        String[] args = new String [] {String.valueOf(id)};
//        String whereClause = String.format(ShiftsEntry._ID + " in (%s)", new Object[] { TextUtils.join(",", Collections.nCopies(args.length, "?")) }); //for deleting multiple lines

        mContext.getContentResolver().delete(ShiftsEntry.CONTENT_URI, ShiftsEntry._ID + "=?", args);
    }

    private String newDate(String dateString){

        String returnString = "01/01/2010";

        String year = dateString.substring(0, 4);
        String month = dateString.substring(5, 7);
        String day = dateString.substring(8);

        returnString = day + "-" + month + "-" + year;

        return returnString;
    }

    public static String[] timeValues(Float duration){

        int hours = (int) Math.floor(duration);
        int minutes = (int) ((duration - hours)*60);

        String hoursString = hours + "";
        String minutesString = String.format("%02d", minutes);

        return new String[]{hoursString,minutesString};
    }

}

File: app/src/main/java/com/appttude/h_mal/farmr/ShiftsCursorAdapter.kt
package com.appttude.h_mal.farmr

import android.app.AlertDialog
import android.content.ContentUris
import android.content.Context
import android.content.DialogInterface
import android.database.Cursor
import android.net.Uri
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.widget.CursorAdapter
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.FragmentTransaction
import com.appttude.h_mal.farmr.data.ShiftProvider
import com.appttude.h_mal.farmr.data.ShiftsContract.ShiftsEntry
import kotlin.math.floor

/**
 * Created by h_mal on 26/12/2017.
 */
class ShiftsCursorAdapter constructor(private val activity: MainActivity, c: Cursor?) : CursorAdapter(activity, c, 0) {
    private var mContext: Context? = null
    var shiftProvider: ShiftProvider? = null
    override fun newView(context: Context, cursor: Cursor, parent: ViewGroup): View {
        return LayoutInflater.from(context).inflate(R.layout.list_item_1, parent, false)
    }

    override fun bindView(view: View, context: Context, cursor: Cursor) {
        mContext = context
        val descriptionTextView: TextView = view.findViewById<View>(R.id.location) as TextView
        val dateTextView: TextView = view.findViewById<View>(R.id.date) as TextView
        val totalPay: TextView = view.findViewById<View>(R.id.total_pay) as TextView
        val hoursView: TextView = view.findViewById<View>(R.id.hours) as TextView
        val h: TextView = view.findViewById<View>(R.id.h) as TextView
        val minutesView: TextView = view.findViewById<View>(R.id.minutes) as TextView
        val m: TextView = view.findViewById<View>(R.id.m) as TextView
        val editView: ImageView = view.findViewById<View>(R.id.imageView) as ImageView
        h.text = "h"
        m.text = "m"
        val typeColumnIndex: String = cursor.getString(cursor.getColumnIndexOrThrow(ShiftsEntry.COLUMN_SHIFT_TYPE))
        val descriptionColumnIndex: String = cursor.getString(cursor.getColumnIndexOrThrow(ShiftsEntry.COLUMN_SHIFT_DESCRIPTION))
        val dateColumnIndex: String = cursor.getString(cursor.getColumnIndexOrThrow(ShiftsEntry.COLUMN_SHIFT_DATE))
        val durationColumnIndex: Float = cursor.getFloat(cursor.getColumnIndexOrThrow(ShiftsEntry.COLUMN_SHIFT_DURATION))
        val unitsColumnIndex: Float = cursor.getFloat(cursor.getColumnIndexOrThrow(ShiftsEntry.COLUMN_SHIFT_UNIT))
        val totalpayColumnIndex: Float = cursor.getFloat(cursor.getColumnIndexOrThrow(ShiftsEntry.COLUMN_SHIFT_TOTALPAY))
        descriptionTextView.text = descriptionColumnIndex
        dateTextView.text = newDate(dateColumnIndex)
        totalPay.text = String.format("%.2f", totalpayColumnIndex)
        if ((typeColumnIndex == "Piece Rate") && durationColumnIndex == 0f) {
            hoursView.text = unitsColumnIndex.toString()
            h.text = ""
            minutesView.text = ""
            m.text = "pcs"
        } else  //            if(typeColumnIndex.equals("Hourly") || typeColumnIndex.equals("hourly"))
        {
            hoursView.text = timeValues(durationColumnIndex).get(0)
            minutesView.text = timeValues(durationColumnIndex).get(1)
        }
        val ID: Long = cursor.getLong(cursor.getColumnIndexOrThrow(ShiftsEntry._ID))
        val currentProductUri: Uri = ContentUris.withAppendedId(ShiftsEntry.CONTENT_URI, ID)
        val b: Bundle = Bundle()
        b.putString("uri", currentProductUri.toString())
        view.setOnClickListener { //                activity.clickOnViewItem(ID);
            val fragmentTransaction: FragmentTransaction = (activity.fragmentManager)!!.beginTransaction()
            val fragment2: FurtherInfoFragment = FurtherInfoFragment()
            fragment2.arguments = b
            fragmentTransaction.replace(R.id.container, fragment2).addToBackStack("furtherinfo").commit()
        }
        editView.setOnClickListener {
            val fragmentTransaction: FragmentTransaction = (activity.fragmentManager)!!.beginTransaction()
            val fragment3: FragmentAddItem = FragmentAddItem()
            fragment3.arguments = b
            fragmentTransaction.replace(R.id.container, fragment3).addToBackStack("additem").commit()
        }
        view.setOnLongClickListener {
            println("long click: $ID")
            val builder: AlertDialog.Builder = AlertDialog.Builder(mContext)
            builder.setMessage("Are you sure you want to delete")
            builder.setPositiveButton("delete") { dialog, id -> deleteProduct(ID) }
            builder.setNegativeButton("cancel") { dialog, id ->
                dialog?.dismiss()
            }
            val alertDialog: AlertDialog = builder.create()
            alertDialog.show()
            true
        }
    }

    private fun deleteProduct(id: Long) {
        val args: Array<String> = arrayOf(id.toString())
        //        String whereClause = String.format(ShiftsEntry._ID + " in (%s)", new Object[] { TextUtils.join(",", Collections.nCopies(args.length, "?")) }); //for deleting multiple lines
        mContext!!.contentResolver.delete(ShiftsEntry.CONTENT_URI, ShiftsEntry._ID + "=?", args)
    }

    private fun newDate(dateString: String): String {
        var returnString: String? = "01/01/2010"
        val year: String = dateString.substring(0, 4)
        val month: String = dateString.substring(5, 7)
        val day: String = dateString.substring(8)
        returnString = "$day-$month-$year"
        return returnString
    }

    companion object {
        fun timeValues(duration: Float): Array<String> {
            val hours: Int = floor(duration.toDouble()).toInt()
            val minutes: Int = ((duration - hours) * 60).toInt()
            val hoursString: String = hours.toString() + ""
            val minutesString: String = String.format("%02d", minutes)
            return arrayOf(hoursString, minutesString)
        }
    }
}